File: tests/upgrade/blueprint_fixture.py
# ─── CGRF Header ───────────────────────────────────────────────
# File:        tests/upgrade/blueprint_fixture.py
# Stage:       08_TEST
# SRS:         SRS-BUILDANDDO-UPGRADE-001
# CAPS:        pending
# CK:          pending
# Dispatch:    VCC-BUILDANDDO-UPGRADE-001
# Seat:        BITS-CODEGEN
# Owner:       Citadel Nexus Inc.
# Created:     2026-09-17
# Depends:     apps/research/blueprints.py
# EnumType:    Test
# EnumEdges:   VALIDATES apps/research/blueprints.py
# DAG Node:    none
# Intent:      Generate a reproducible text-only specification PDF without adding a PDF creation dependency.
# ───────────────────────────────────────────────────────────────

"""Build a small PDF fixture using the PDF text operators consumed by pypdf."""
from __future__ import annotations

from pathlib import Path

SPEC = """CUSTOMER PORTAL BLUEPRINT
1. Architecture
API Gateway connects to Account Service and reads from User Database.
Account Service calls Mail Worker.
Mail Worker writes to Event Queue.
2. Functional requirements
REQ-001: Account Service shall register new accounts.
REQ-002: API Gateway should respond within 200 milliseconds.
REQ-003: User Database must encrypt personal data.
REQ-004: Mail Worker could send weekly summaries.
REQ-005: The system won't support anonymous purchases.
3. Constraints
The monthly budget must not exceed 500 dollars.
4. Assumptions
Assume the customer supplies an identity provider.
5. Open questions
TBD: data retention policy.
The deployment region is not yet decided.
"""


def pdf_bytes(value: str = SPEC, *, pages: int = 1, active_content: bool = False) -> bytes:
    """Build valid fixture bytes with optional inert PDF JavaScript metadata."""
    content = "BT /F1 11 Tf 45 780 Td 15 TL\n" + "\n".join(
        "(" + line.replace("\\", "\\\\").replace("(", "\\(").replace(")", "\\)") + ") Tj T*"
        for line in value.splitlines()
    ) + "\nET"
    stream = content.encode("latin-1")
    objects = [
        b"<< /Type /Catalog /Pages 2 0 R" + (b" /OpenAction << /S /JavaScript /JS (throw 'untrusted';) >>" if active_content else b"") + b" >>",
        ("<< /Type /Pages /Kids [" + " ".join(f"{5 + i} 0 R" for i in range(pages)) + f"] /Count {pages} >>").encode(),
        b"<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>",
        b"<< /Length " + str(len(stream)).encode() + b" >>\nstream\n" + stream + b"\nendstream",
    ]
    objects.extend(b"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 840] /Resources << /Font << /F1 3 0 R >> >> /Contents 4 0 R >>" for _ in range(pages))
    # A binary comment prevents Git from normalizing the fixture's byte offsets.
    data = bytearray(b"%PDF-1.4\n%\xe2\xe3\xcf\xd3\x00\n%BuildAndDo synthetic fixture\n")
    offsets = [0]
    for number, body in enumerate(objects, 1):
        offsets.append(len(data))
        data.extend(f"{number} 0 obj\n".encode() + body + b"\nendobj\n")
    xref = len(data)
    data.extend(f"xref\n0 {len(objects) + 1}\n0000000000 65535 f \n".encode())
    for offset in offsets[1:]:
        data.extend(f"{offset:010} 00000 n \n".encode())
    data.extend(f"trailer\n<< /Size {len(objects) + 1} /Root 1 0 R >>\nstartxref\n{xref}\n%%EOF\n".encode())
    return bytes(data)


if __name__ == "__main__":
    target = Path(__file__).with_name("fixtures") / "sample-blueprint.pdf"
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_bytes(pdf_bytes())

File: tests/upgrade/blueprint-fixture.mjs
// ─── CGRF Header ───────────────────────────────────────────────
// File:        tests/upgrade/blueprint-fixture.mjs
// Stage:       08_TEST
// SRS:         SRS-BUILDANDDO-UPGRADE-001
// CAPS:        pending
// CK:          pending
// Dispatch:    VCC-BUILDANDDO-UPGRADE-001
// Seat:        BITS-CODEGEN
// Owner:       Citadel Nexus Inc.
// Created:     2026-09-17
// Depends:     apps/pocketbase/pb_hooks/workspace-blueprints.js
// EnumType:    Test
// EnumEdges:   VALIDATES apps/pocketbase/pb_hooks/workspace-blueprints.js
// DAG Node:    none
// Intent:      Reuse the transactional research fixture for workspace blueprint intake and worker completion.
// ───────────────────────────────────────────────────────────────

import { createHash } from 'node:crypto';
import { researchFixture } from './research-fixture.mjs';
import { plain } from './admin-fixture.mjs';

export const SCHEMA = 'apps/pocketbase/pb_migrations/1790500000_workspace_blueprints.js';
export const hash = (value) => createHash('sha256').update(value).digest('hex');
export function blueprintFixture() {
    const f = researchFixture({ runtime: { $security: { sha256: hash } } });
    f.migration(SCHEMA).up();
    const blueprint = f.load('workspace-blueprints.js');
    let sequence = 0;
    const upload = ({ actor = 'editor', workspace = 'ws1', name = 'sample.pdf', bytes = Buffer.from('%PDF-fixture'), key,
        fields = {}, count = 1, size = bytes.length } = {}) => {
        const body = { request_key: key || 'blueprint_request_' + ++sequence, input_sha256: hash(bytes), ...fields };
        const event = f.event(actor, body, { workspace });
        event.findUploadedFiles = () => Array.from({ length: count }, () => ({ originalName: name, name, size }));
        const result = plain(blueprint.upload(event));
        const submission = f.app.findRecordById('research_submissions', result.record.submission);
        // Model only native file persistence; all access and application handlers run.
        const file = f.app.findRecordById('research_uploads', submission.getString('upload'));
        file.set('asset', name); f.app.save(file);
        return result;
    };
    const result = {
        text: 'Account Service must encrypt records.', citations: [], processor: 'local-blueprint', version: 'pypdf-fixture',
        input_sha256: hash(Buffer.from('%PDF-fixture')), truncated: false, blueprint_failure: '', evaluation_failure: '', evaluation: null,
        blueprint: { title: 'Portal', source_file: 'sample.pdf', source_hash: hash(Buffer.from('%PDF-fixture')),
            extracted_at: '2026-09-17T00:00:00+00:00', parser_version: 'blueprint-1.0', page_count: 1, truncated: false, extraction_confidence: .7,
            sections: [{ id: 'SEC-001', title: 'Security', text: 'Account Service must encrypt records.', level: 1, start_line: 1, end_line: 2 }],
            requirements: [{ id: 'REQ-001', text: 'Account Service must encrypt records.', priority: 'must', type: 'security',
                section: 'SEC-001', raw_context: 'Account Service must encrypt records.' }],
            components: [{ name: 'Account Service', type: 'service', description: 'Account Service must encrypt records.', requirements: ['REQ-001'], dependencies: [] }],
            constraints: [], assumptions: [], open_questions: [] },
    };
    const questions = ['feasibility', 'complexity', 'component_type', 'automatable', 'risk'];
    result.evaluation = {
        source_hash: result.input_sha256, authority: 'A0', verified: false,
        requirements: [{ requirement_id: 'REQ-001', feasibility: null, complexity: null, component_type: null, automatable: null, risk: null,
            confidence: Object.fromEntries(questions.map((key) => [key, .5])), abstained: questions, components: ['Account Service'],
            route: 'abstain', trace_id: 'test-trace', state_hash: 'a'.repeat(64) }],
        overall: { status: 'review_required', requirement_count: 1, assessed_count: 0, review_count: 1,
            average_scores: { feasibility: null, complexity: null, risk: null } },
    };
    const detail = (id, actor = 'editor', workspace = 'ws1') => plain(blueprint.detail(f.event(actor, {}, { id, workspace })));
    const command = (id, action, revision, actor = 'editor', key = 'blueprint_command_' + ++sequence) =>
        plain(blueprint.command(f.event(actor, { action, revision, request_key: key }, { id })));
    return { ...f, get data() { return f.data; }, blueprint, upload, detail, blueprintCommand: command, result };
}
